nupic/embodied/envs/wrappers.py

```diff
@@ -0,0 +1,539 @@
+import itertools
+from collections import deque
+from copy import copy
+
+import gym
+import numpy as np
+from PIL import Image
+import pybullet
+
+
+def unwrap(env):
+    if hasattr(env, "unwrapped"):
+        return env.unwrapped
+    elif hasattr(env, "env"):
+        return unwrap(env.env)
+    elif hasattr(env, "leg_env"):
+        return unwrap(env.leg_env)
+    else:
+        return env
+
+
+class MaxAndSkipEnv(gym.Wrapper):
+    def __init__(self, env, skip=4):
+        """Return only every `skip`-th frame"""
+        gym.Wrapper.__init__(self, env)
+        # most recent raw observations (for max pooling across time steps)
+        self._obs_buffer = deque(maxlen=2)
+        self._skip = skip
+
+    def step(self, action):
+        """Repeat action, sum reward, and max over last observations."""
+        total_reward = 0.0
+        done = None
+        acc_info = {}
+        for _ in range(self._skip):
+            obs, reward, done, info = self.env.step(action)
+            acc_info.update(info)
+            self._obs_buffer.append(obs)
+            total_reward += reward
+            if done:
+                break
+        max_frame = np.max(np.stack(self._obs_buffer), axis=0)
+
+        return max_frame, total_reward, done, acc_info
+
+    def reset(self):
+        """Clear past frame buffer and init. to first obs. from inner env."""
+        self._obs_buffer.clear()
+        obs = self.env.reset()
+        self._obs_buffer.append(obs)
+        return obs
+
+
+class ProcessFrame84(gym.ObservationWrapper):
+    def __init__(self, env, crop=True):
+        self.crop = crop
+        super(ProcessFrame84, self).__init__(env)
+        self.observation_space = gym.spaces.Box(
+            low=0, high=255, shape=(84, 84, 1), dtype=np.uint8
+        )
+
+    def observation(self, obs):
+        return ProcessFrame84.process(obs, crop=self.crop)
+
+    @staticmethod
+    def process(frame, crop=True):
+        if frame.size == 210 * 160 * 3:
+            img = np.reshape(frame, [210, 160, 3]).astype(np.float32)
+        elif frame.size == 250 * 160 * 3:
+            img = np.reshape(frame, [250, 160, 3]).astype(np.float32)
+        elif frame.size == 224 * 240 * 3:  # mario resolution
+            img = np.reshape(frame, [224, 240, 3]).astype(np.float32)
+        else:
+            assert False, "Unknown resolution." + str(frame.size)
+        img = img[:, :, 0] * 0.299 + img[:, :, 1] * 0.587 + img[:, :, 2] * 0.114
+        size = (84, 110 if crop else 84)
+        resized_screen = np.array(
+            Image.fromarray(img).resize(size, resample=Image.BILINEAR), dtype=np.uint8
+        )
+        x_t = resized_screen[18:102, :] if crop else resized_screen
```

Comment:
It would be good to have some more background on what do this constants represent. We could either assign them to variables with a meaningful descriptive name or add a comment instead if assigning them to a name doesn't make sense.